include/deal.II/matrix_free/assemble_operator.h

```diff
@@ -0,0 +1,278 @@
+     * This is useful if the matrix does not change its pattern.
+     * // TODO cache entries
+     */
+    struct GraphCache
+    {
+      Graph                           graph;
+      std::vector<vertices_size_type> color_vec;
+      boost::iterator_property_map<vertices_size_type *, vertex_index_map>
+                                                   coloring;
+      vertices_size_type                           num_colors;
+      std::vector<std::vector<vertices_size_type>> entries_per_color;
+      unsigned int                                 max_entries_per_row = 0;
+
+      /**
+       * Create a boost graph (V,E) from the provided sparsity pattern.
+       * The vertices V are given by the column indices.
+       * An edge (v1,v2) is an element of E, if there is a row r,
+       * such that (r,v1) and (r,v2) are both entries of the sparsity pattern.
+       *
+       * @param <tt>pattern</tt> the sparsity pattern for which the graph should be built.
+       * Pattern needs to provide: n_rows(), n_cols(), begin(r), end(r),
+       * iterator->row(), iterator->colum().
+       */
+      template <typename Pattern>
+      void
+      make_graph(const Pattern &pattern)
+      {
+        std::set<Edge> edges;
+        for (unsigned int r = 0; r < pattern.n_rows(); r++)
+          {
+            for (auto it = pattern.begin(r); it != pattern.end(r); it++)
+              {
+                for (auto it2 = it; it2 != pattern.end(r); it2++)
+                  {
+                    AssertDimension(r, it->row());
+                    AssertDimension(r, it2->row());
+                    auto c1 = it->column();
+                    auto c2 = it2->column();
+                    if (c1 > c2)
+                      std::swap(c1, c2);
+                    edges.insert(std::make_pair(c1, c2));
+                  }
+              }
+          }
+
+        // Construct a graph from the edge-set.
+        for (const Edge &e : edges)
+          add_edge(e.first, e.second, graph);
+
+        AssertDimension(pattern.n_cols(), boost::num_vertices(graph));
+
+        max_entries_per_row = pattern.max_entries_per_row();
+      }
+
+      /**
+       * Color the existing graph using sequential vertex coloring.
+       * Different orderings of the vertices can give different results.
+       * By default, we do not reorder them, as the benefit does not seem to be
+       * huge.
+       *
+       * @param <tt>order</tt> defines a reording of the vertices.
+       * The given type needs to satisfy the boost::property_map concept.
+       * Example:
+       *   boost::vector_property_map<vertices_size_type> order;
+       *   boost::smallest_last_vertex_ordering(graph, order);
+       */
+      template <
+        typename Order = boost::typed_identity_property_map<vertices_size_type>>
+      void
+      make_coloring(const Order &&order =
+                      boost::typed_identity_property_map<vertices_size_type>{})
```

Comment:
It's not illegal to have `const` rvalue references -- but they are pretty pointless :-) (I had to look this up, because I was confused. But it was a fun read: https://www.codesynthesis.com/~boris/blog/2012/07/24/const-rvalue-references/)

Reply:
changed to const &